src/get-started/write-and-run-unit-tests.md

```diff
@@ -1,10 +1,229 @@
+
+### Running tests for pull request
+
+Before merging a pull request, it requires to pass all tests.
+
+Generally continuous integration will run tests for you, but if you want to run tests with conditions or rerun tests on failure, you should know how to achieve that.
+
+#### `/rebuild`
+
+Build the binary based on the PR for testing. It also reruns all the CI test cases.
+
+#### `/run-all-tests`
+
+Rerun all the CI test cases. This command accepts the following parameters:
+
+* `tidb=<branch>|<pr/$num>` specifies which tidb to use
+* `tikv=<branch>|<pr/$num>` specifies which tikv to use
+* `pd=<branch>|<pr/$num>` specifies which pd to use
+* `tidb-test=<branch>|<pr/$num>` specifies which tidb-test to use
+* `tidb-private-test=<branch>|<pr/$num>` specifies which tidb-private-test to use
+
+for examples,
+
+```
+/run-all-tests tidb-test=pr/666
+/run-all-tests tidb-test=release-2.1
+/run-all-tests tikv=pr/999
+```
+
+#### `/run-{{test}}`
+
+Run a single CI test. This command accepts the same parameters as `/run-all-tests`.
+
+## How to find the failed tests?
+
+There are several common cause of failed tests.
+
+### Assertion failed
+
+The most common cause of failed tests is assertion failed. Its failure report looks like:
+
+```
+=== RUN   TestTopology
+    info_test.go:72: 
+        	Error Trace:	info_test.go:72
+        	Error:      	Not equal: 
+        	            	expected: 1282967700000
+        	            	actual  : 1628585893
+        	Test:       	TestTopology
+--- FAIL: TestTopology (0.76s)
+```
+
+You may `grep "FAIL:"` the output to find this kind of failure.
+
+### Data race
+
+Golang testing supports detecting data race by running tests with `-race` flag. Its failure report looks like:
+
+```
+[2021-06-21T15:36:38.766Z] ==================
+[2021-06-21T15:36:38.766Z] WARNING: DATA RACE
+[2021-06-21T15:36:38.766Z] Read at 0x00c0055ce380 by goroutine 108:
+...
+[2021-06-21T15:36:38.766Z] Previous write at 0x00c0055ce380 by goroutine 169:
+[2021-06-21T15:36:38.766Z]   [failed to restore the stack]
+```
+
+### Goroutine leak
+
+We use goleak to detect goroutine leak for tests. Its failure report looks like:
+
+```
+goleak: Errors on successful test run: found unexpected goroutines:
+[Goroutine 104 in state chan receive, with go.etcd.io/etcd/pkg/logutil.(*MergeLogger).outputLoop on top of the stack:
+goroutine 104 [chan receive]:
+go.etcd.io/etcd/pkg/logutil.(*MergeLogger).outputLoop(0xc000197398)
+	/go/pkg/mod/go.etcd.io/etcd@v0.5.0-alpha.5.0.20200824191128-ae9734ed278b/pkg/logutil/merge_logger.go:173 +0x3ac
+created by go.etcd.io/etcd/pkg/logutil.NewMergeLogger
+	/go/pkg/mod/go.etcd.io/etcd@v0.5.0-alpha.5.0.20200824191128-ae9734ed278b/pkg/logutil/merge_logger.go:91 +0x85
+
+```
+
+You can determine source of package leaks follow the [documentation](https://github.com/uber-go/goleak/#determine-source-of-package-leaks)
+
+### Timeout
```

Comment:
Might be good to add some advise here, e.g:

- Split up large tests
- Using profiling and/or logging to identify where it is spending time
- Move initialization from tests to a separate setup (Link to TestMain)

Reply:
```
go test -check.f '\.TestTLS' -cpuprofile cpu.pro
go tool pprof -show_from='TestTLSAuto$' -web cpu.prof
```

Reply:
I'd prefer you create an issue and do it as a followup. I tend to show the common failure here and we can later elaborate how to overcome them. Or on benchmarking topics, I think "Debug and Profile" section is far from completed.

Resolved as won't do in this PR.